### PR TITLE
KAFKA-9966: add internal assignment listener to stabilize eos-beta upgrade test

### PR DESCRIPTION
**Pull request by ableegoldman**
Adds an internal assignment listener interface with a callback that exposes the assignment stability. A useful implementation that tracks the number of stable assignments is available in IntegrationTestUtils,  and leveraged to fix the flaky EOSBetaUpgradeTest, but any implementation can be plugged in to meet the specific needs of a test.

### CODE REVIEW

**Comment by ableegoldman** [comment]
Will need to be rebased once [pull/8596](https://github.com/apache/kafka/pull/8596/) is merged, but the PR is ready. Reduced the nearly 100% failure rate of this test on top of the changes in [pull/8596](https://github.com/apache/kafka/pull/8596/) to 0% (that I've seen so far 🙂) @mjsax @vvcephei 

edit: this PR alone does not contain all the fixes needed for the eos-beta upgrade test, so I'm leaving it ignored for now. The other set of fixes are included in https://github.com/apache/kafka/pull/8662

We can remove the Ignore annotation in whichever PR gets merged last

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java [hunk 1]
```diff
@@ -185,24 +187,25 @@ public String toString() {
     public void configure(final Map<String, ?> configs) {
         final AssignorConfiguration assignorConfiguration = new AssignorConfiguration(configs);
 
-        logPrefix = assignorConfiguration.logPrefix();
+        logPrefix = assignorConfiguration.getLogPrefix();
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Why do we use `getLogPrefix()` -- our code style _always_ omit the `get` prefix -- should be just `logPrefix()` (Similar below for other methods)

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: I guess I thought there was a reason `getXX` was used in this class and figured all the methods should conform. Idk. I'll remove the `get`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignorConfiguration.java [hunk 2]
```diff
@@ -340,6 +319,27 @@ public TaskAssignor getTaskAssignor() {
         }
     }
 
+    public AssignmentListener assignmentListener() {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignorConfiguration.java:322]: This is the only non-cosmetic change in this class, along with the interface added below

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignorConfiguration.java [hunk 3]
```diff
@@ -47,59 +47,21 @@
 
     private final String logPrefix;
     private final Logger log;
-    private final AssignmentConfigs assignmentConfigs;
-    @SuppressWarnings("deprecation")
-    private final org.apache.kafka.streams.processor.PartitionGrouper partitionGrouper;
-    private final String userEndPoint;
     private final TaskManager taskManager;
-    private final StreamsMetadataState streamsMetadataState;
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignorConfiguration.java:55]: Sorry for all the lines of changes here, but the inconsistent method signatures and object construction was making this class hard to follow. All I'm doing here is making the methods conform to the same style, and moving the construction of any config that isn't needed elsewhere to its getter method. This allows us to remove a lot of these class variables

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java [hunk 4]
```diff
@@ -906,29 +912,22 @@ public void close() { }
         return streams;
     }
 
-    private void waitForStateTransition(final List<KeyValue<KafkaStreams.State, KafkaStreams.State>> observed,
-                                        final List<KeyValue<KafkaStreams.State, KafkaStreams.State>> expected)
-            throws Exception {
-
+    private void waitForRunning(final List<KeyValue<KafkaStreams.State, KafkaStreams.State>> observed) throws Exception {
```
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:914]: This one actually might not be necessary, but I thought it was useful as a sanity check especially since we're trying to access the stores in some places

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignorConfiguration.java (hunk 3 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignorConfiguration.java:55]: Do we know if the corresponding "getters" are called often or not? I guess the main idea behind having those member variables was to "parse" the config once as it's immutable anyway instead of each time a getter is called?

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java (hunk 4 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java:914]: Agreed.

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java [hunk 5]
```diff
@@ -906,29 +912,22 @@ public void close() { }
...
         waitForCondition(
-            () -> observed.equals(expected),
+            () -> observed.get(observed.size() - 1).equals(new KeyValue<>(State.REBALANCING, State.RUNNING)),
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java]: Seems this could fail if `observed` is empty?

Also, could we simplify this? We are only interested in `RUNNING ` state, right? So no need to check the previous state?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java [hunk 6]
```diff
@@ -1274,6 +1274,7 @@ public void shouldThrowExceptionIfApplicationServerConfigIsNotHostPortPair() {
 
     @Test
     public void shouldThrowExceptionIfApplicationServerConfigPortIsNotAnInteger() {
+        createDefaultMockTaskManager();
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java:1277]: Why this change?

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java [hunk 7]
```diff
@@ -261,24 +241,24 @@ public void shouldUpgradeFromEosAlphaToEosBeta() throws Exception {
...
             streams1Alpha.cleanUp();
             streams1Alpha.start();
-            waitForStateTransition(
-                stateTransitions1,
-                Arrays.asList(
-                    KeyValue.pair(KafkaStreams.State.CREATED, KafkaStreams.State.REBALANCING),
-                    KeyValue.pair(KafkaStreams.State.REBALANCING, KafkaStreams.State.RUNNING)
-                )
-            );
+            assignmentListener.waitForNextStableAssignment(expectedNumStableAssignments, MAX_WAIT_TIME_MS);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java]: So far we _always_ pass in `assignmentListener.numStableAssignments() + 1` -- hence, I am wondering if we need to pass it at all?

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java [hunk 8]
```diff
@@ -466,6 +452,9 @@ public void shouldUpgradeFromEosAlphaToEosBeta() throws Exception {
                 verifyCommitted(expectedCommittedResult);
                 expectedUncommittedResult.addAll(expectedCommittedResult);
             } else {
+                waitForRunning(stateTransitions1);
+                waitForRunning(stateTransitions2);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java]: Why do we need this? If we don't inject an error, we start phase 6 in RUNNING state and it should not rebalance.

**Review by mjsax — COMMENTED** [review]
Overall LGTM.

**Comment by mjsax** [comment]
Retest this please.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignorConfiguration.java (hunk 3 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignorConfiguration.java:55]: They are not. We construct the `AssignorConfiguration` object once in `StreamsPartitionAssignor#configure` and call the getters immediately afterwards, and only there

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java (hunk 7 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java]: Well we need to record it _before_ doing whatever action causes the group to rebalance. But maybe we can do something like `prepareForRebalance` if that seems easier?

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java (hunk 8 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java]: Good point

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java (hunk 5 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java]: True, I'll fix this

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java (hunk 6 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java:1277]: If we don't configure the TaskManager then we'll throw (the wrong kind of) exception during `configure`. Previously this test just happened to pass because we processed the user endpoint before we processed the TaskManager, so we never got to the TaskManager

File: streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java (hunk 7 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java]: I see. Guess it's good as-is.

**Review by mjsax — APPROVED** [review]
LGTM.

**Comment by mjsax** [comment]
Retest this please.

**Comment by mjsax** [comment]
Retest this please.

**Comment by ableegoldman** [comment]
Java 14 passed, Java 8 build failed due to mysterious `17:27:43 ERROR: Error cloning remote repo 'origin'` and Java 11 failed due to flaky `MirrorConnectorsIntegrationTest.testReplication` (reported on ticket)
